cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -0,0 +1,1100 @@
+      return Status::OK();
+    }
+    // Not found => perhaps it's a non-empty "directory"
+    RETURN_NOT_OK(impl_->IsNonEmptyDirectory(path, &is_dir));
+    if (is_dir) {
+      st.set_type(FileType::Directory);
+    } else {
+      st.set_type(FileType::NonExistent);
+    }
+    *out = st;
+    return Status::OK();
+  }
+}
+
+Status S3FileSystem::GetTargetStats(const Selector& select, std::vector<FileStats>* out) {
+  S3Path base_path;
+  RETURN_NOT_OK(S3Path::FromString(select.base_dir, &base_path));
+  out->clear();
+
+  if (base_path.empty()) {
+    // List all buckets
+    std::vector<std::string> buckets;
+    RETURN_NOT_OK(impl_->ListBuckets(&buckets));
+    for (const auto& bucket : buckets) {
+      FileStats st;
+      st.set_path(bucket);
+      st.set_type(FileType::Directory);
+      out->push_back(std::move(st));
+      if (select.recursive) {
+        RETURN_NOT_OK(impl_->Walk(select, bucket, "", out));
+      }
+    }
+    return Status::OK();
+  }
+
+  // Nominal case -> walk a single bucket
+  return impl_->Walk(select, base_path.bucket, base_path.key, out);
+}
+
+Status S3FileSystem::CreateDir(const std::string& s, bool recursive) {
+  S3Path path;
+  RETURN_NOT_OK(S3Path::FromString(s, &path));
+
+  if (path.key.empty()) {
+    // Create bucket
+    return impl_->CreateBucket(path.bucket);
+  }
+
+  // Create object
+  if (recursive) {
+    // Ensure bucket exists
+    RETURN_NOT_OK(impl_->CreateBucket(path.bucket));
+    // Ensure that all parents exist, then the directory itself
+    std::string parent_key;
+    for (auto it = path.key_parts.begin(); it != path.key_parts.end(); ++it) {
+      parent_key += *it;
+      parent_key += kSep;
+      RETURN_NOT_OK(impl_->CreateEmptyObject(path.bucket, parent_key));
+    }
+    return Status::OK();
+  } else {
+    // Check parent dir exists
+    if (path.has_parent()) {
+      S3Path parent_path = path.parent();
+      bool exists;
+      RETURN_NOT_OK(impl_->IsNonEmptyDirectory(parent_path, &exists));
+      if (!exists) {
+        RETURN_NOT_OK(impl_->IsEmptyDirectory(parent_path, &exists));
+      }
+      if (!exists) {
+        return Status::IOError("Cannot create directory '", path.full_path,
+                               "': parent directory does not exist");
+      }
+    }
+
+    // XXX Should we check that no non-directory entry exists?
+    // Minio does it for us, not sure about other S3 implementations.
+    return impl_->CreateEmptyDir(path.bucket, path.key);
+  }
+}
+
+Status S3FileSystem::DeleteDir(const std::string& s) {
+  S3Path path;
+  RETURN_NOT_OK(S3Path::FromString(s, &path));
+
+  if (path.empty()) {
+    return Status::NotImplemented("Cannot delete all S3 buckets");
+  }
+  RETURN_NOT_OK(impl_->DeleteDir(path.bucket, path.key));
+  if (path.key.empty()) {
+    // Also delete bucket
+    S3Model::DeleteBucketRequest req;
+    req.SetBucket(ToAwsString(path.bucket));
+    return OutcomeToStatus(impl_->client_->DeleteBucket(req));
+  }
```

Comment:
Unclear if we want to allow this function to delete buckets (since these are the "heaviest" of S3 objects) by default. It might be better to error when trying to delete a bucket and instead offset a `DeleteBucket` API

Reply:
At this point, the bucket is already empty, though...